File: platform/pkg/observability/logger/noop.go
// platform/pkg/observability/logger/noop.go
package logger

import (
	"context"
	"fmt"
	"strings"
	"time"
)

// ANSI color codes
const (
	colorReset  = "\033[0m"
	colorRed    = "\033[31m"
	colorGreen  = "\033[32m"
	colorYellow = "\033[33m"
	colorBlue   = "\033[34m"
	colorPurple = "\033[35m"
	colorCyan   = "\033[36m"
	colorGray   = "\033[37m"
	colorWhite  = "\033[97m"
)

// noopLogger is a simple console logger with colored output.
// Useful for local development and debugging.
type noopLogger struct {
	fields []any
}

// NewNoop creates a new console logger with colored output.
// Great for local development - logs are human-readable with colors.
func NewNoop() Logger {
	return &noopLogger{}
}

func (l *noopLogger) Debug(msg string, keysAndValues ...any) {
	l.log(DebugLevel, msg, keysAndValues...)
}

func (l *noopLogger) Info(msg string, keysAndValues ...any) {
	l.log(InfoLevel, msg, keysAndValues...)
}

func (l *noopLogger) Warn(msg string, keysAndValues ...any) {
	l.log(WarnLevel, msg, keysAndValues...)
}

func (l *noopLogger) Error(msg string, keysAndValues ...any) {
	l.log(ErrorLevel, msg, keysAndValues...)
}

func (l *noopLogger) With(keysAndValues ...any) Logger {
	newFields := make([]any, len(l.fields)+len(keysAndValues))
	copy(newFields, l.fields)
	copy(newFields[len(l.fields):], keysAndValues)
	return &noopLogger{fields: newFields}
}

func (l *noopLogger) WithContext(ctx context.Context) Logger {
	// Could extract values from context if needed
	return l
}

func (l *noopLogger) WithError(err error) Logger {
	return l.With("error", err.Error())
}

// log outputs a colored log message to stdout
func (l *noopLogger) log(level Level, msg string, keysAndValues ...any) {
	// Timestamp
	timestamp := time.Now().Format("15:04:05.000")
	
	// Level with color
	levelStr := l.colorizeLevel(level)
	
	// Message
	msgStr := fmt.Sprintf("%s%s%s", colorWhite, msg, colorReset)
	
	// Combine persistent fields with new fields
	allFields := append(l.fields, keysAndValues...)
	
	// Format fields
	fieldsStr := l.formatFields(allFields)
	
	// Output
	if fieldsStr != "" {
		fmt.Printf("%s%s%s %s %s %s\n",
			colorGray, timestamp, colorReset,
			levelStr,
			msgStr,
			fieldsStr,
		)
	} else {
		fmt.Printf("%s%s%s %s %s\n",
			colorGray, timestamp, colorReset,
			levelStr,
			msgStr,
		)
	}
}

// colorizeLevel returns a colored level string
func (l *noopLogger) colorizeLevel(level Level) string {
	switch level {
	case DebugLevel:
		return fmt.Sprintf("%sDEBUG%s", colorCyan, colorReset)
	case InfoLevel:
		return fmt.Sprintf("%sINFO %s", colorGreen, colorReset)
	case WarnLevel:
		return fmt.Sprintf("%sWARN %s", colorYellow, colorReset)
	case ErrorLevel:
		return fmt.Sprintf("%sERROR%s", colorRed, colorReset)
	default:
		return "UNKNOWN"
	}
}

// formatFields formats key-value pairs with colors
func (l *noopLogger) formatFields(keysAndValues []any) string {
	if len(keysAndValues) == 0 {
		return ""
	}
	
	var parts []string
	for i := 0; i < len(keysAndValues); i += 2 {
		if i+1 >= len(keysAndValues) {
			break
		}
		
		key := keysAndValues[i]
		value := keysAndValues[i+1]
		
		// Color the key
		keyStr := fmt.Sprintf("%s%v%s", colorPurple, key, colorReset)
		
		// Format the value
		valueStr := fmt.Sprintf("%v", value)
		
		parts = append(parts, fmt.Sprintf("%s=%s", keyStr, valueStr))
	}
	
	return strings.Join(parts, " ")
}

File: platform/pylib/scout_common/observability/logging/noop.py
"""Console logger with colored output for local development."""

from typing import Any
from datetime import datetime
import sys


# ANSI color codes
class Colors:
    RESET = "\033[0m"
    RED = "\033[31m"
    GREEN = "\033[32m"
    YELLOW = "\033[33m"
    BLUE = "\033[34m"
    PURPLE = "\033[35m"
    CYAN = "\033[36m"
    GRAY = "\033[37m"
    WHITE = "\033[97m"


class NoopLogger:
    """
    Simple console logger with colored output.
    
    Great for local development - logs are human-readable with colors.
    Not a true "no-op" but a simple stdout logger without structured backends.
    """
    
    def __init__(self, **fields: Any) -> None:
        """
        Initialize console logger.
        
        Args:
            **fields: Persistent fields to attach to all log messages
        """
        self._fields = fields
    
    def debug(self, msg: str, **kwargs: Any) -> None:
        """Log debug message with cyan color."""
        self._log("DEBUG", Colors.CYAN, msg, **kwargs)
    
    def info(self, msg: str, **kwargs: Any) -> None:
        """Log info message with green color."""
        self._log("INFO ", Colors.GREEN, msg, **kwargs)
    
    def warn(self, msg: str, **kwargs: Any) -> None:
        """Log warning message with yellow color."""
        self._log("WARN ", Colors.YELLOW, msg, **kwargs)
    
    def error(self, msg: str, **kwargs: Any) -> None:
        """Log error message with red color."""
        self._log("ERROR", Colors.RED, msg, **kwargs)
    
    def with_fields(self, **kwargs: Any) -> "NoopLogger":
        """
        Return a new logger with fields attached.
        
        Args:
            **kwargs: Fields to attach
            
        Returns:
            New logger instance with combined fields
        """
        new_fields = {**self._fields, **kwargs}
        return NoopLogger(**new_fields)
    
    def with_error(self, err: Exception) -> "NoopLogger":
        """
        Return a new logger with error attached.
        
        Args:
            err: Exception to attach
            
        Returns:
            New logger instance with error fields
        """
        return self.with_fields(
            error=str(err),
            error_type=type(err).__name__,
        )
    
    def _log(self, level: str, color: str, msg: str, **kwargs: Any) -> None:
        """
        Output a colored log message to stdout.
        
        Args:
            level: Log level string
            color: ANSI color code
            msg: Log message
            **kwargs: Additional fields
        """
        # Timestamp
        timestamp = datetime.now().strftime("%H:%M:%S.%f")[:-3]
        
        # Combine persistent fields with new fields
        all_fields = {**self._fields, **kwargs}
        
        # Format fields
        fields_str = self._format_fields(all_fields)
        
        # Build output
        parts = [
            f"{Colors.GRAY}{timestamp}{Colors.RESET}",
            f"{color}{level}{Colors.RESET}",
            f"{Colors.WHITE}{msg}{Colors.RESET}",
        ]
        
        if fields_str:
            parts.append(fields_str)
        
        # Print to stdout
        print(" ".join(parts), file=sys.stdout)
        sys.stdout.flush()
    
    def _format_fields(self, fields: dict[str, Any]) -> str:
        """
        Format key-value pairs with colors.
        
        Args:
            fields: Dictionary of fields
            
        Returns:
            Formatted string with colored keys
        """
        if not fields:
            return ""
        
        parts = []
        for key, value in fields.items():
            key_str = f"{Colors.PURPLE}{key}{Colors.RESET}"
            parts.append(f"{key_str}={value}")
        
        return " ".join(parts)